mc/query/stmt.go

```diff
@@ -9,13 +9,72 @@ func StatementRefs(stmt *pb.Statement) []string {
 	case *pb.StatementBody_Simple:
 		return body.Simple.Refs
 
+	case *pb.StatementBody_Compound:
+		stmts := body.Compound.Body
+		count := 0
+		for _, stmt := range stmts {
+			count += len(stmt.Refs)
+		}
+		refs := make([]string, 0, count)
+		for _, stmt := range stmts {
+			refs = append(refs, stmt.Refs...)
+		}
+		return refs
+
+	case *pb.StatementBody_Envelope:
+		stmts := body.Envelope.Body
+		count := 0
+		for _, stmt := range stmts {
+			count += countStatementRefs(stmt)
+		}
+		refs := make([]string, 0, count)
+		for _, stmt := range stmts {
+			refs = append(refs, StatementRefs(stmt)...)
+		}
+		return refs
+
 	default:
-		// TODO compound statements etc
 		return nil
 	}
 }
 
+func countStatementRefs(stmt *pb.Statement) int {
+	switch body := stmt.Body.Body.(type) {
+	case *pb.StatementBody_Simple:
+		return len(body.Simple.Refs)
+
+	case *pb.StatementBody_Compound:
+		stmts := body.Compound.Body
+		count := 0
+		for _, stmt := range stmts {
+			count += len(stmt.Refs)
+		}
+		return count
+
+	case *pb.StatementBody_Envelope:
+		stmts := body.Envelope.Body
+		count := 0
+		for _, stmt := range stmts {
+			count += countStatementRefs(stmt)
+		}
+		return count
+
+	default:
+		return 0
+	}
+}
+
 func StatementSource(stmt *pb.Statement) string {
-	// TODO compound statements etc
-	return stmt.Publisher
+	switch body := stmt.Body.Body.(type) {
+	case *pb.StatementBody_Envelope:
+		// that's just the source of the first statement; all envelope statements
+		// should have the same source
+		if len(body.Envelope.Body) > 0 {
+			return StatementSource(body.Envelope.Body[0])
```

Comment:
Yeah, I think we want to consider it a constraint.
In practical terms, it requires too many changes (including breaking db schema change) and introduces too much complexity to have multiple sources per statement.
So let's make it a decree -- statements have a single origin source.